src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpGoToImplementation.cs

```diff
@@ -40,9 +44,35 @@ public async Task SimpleGoToImplementation()
 }", HangMitigatingCancellationToken);
             await TestServices.Editor.PlaceCaretAsync("interface IGoo", charsOffset: 0, HangMitigatingCancellationToken);
             await TestServices.Editor.GoToImplementationAsync(HangMitigatingCancellationToken);
+
+            string identifierWithCaret;
+            var activeCaption = await TestServices.Shell.GetActiveWindowCaptionAsync(HangMitigatingCancellationToken);
             var dirtyModifier = await TestServices.Editor.GetDirtyIndicatorAsync(HangMitigatingCancellationToken);
+            if (activeCaption == $"FileImplementation.cs{dirtyModifier}")
+            {
+                // The navigation completed synchronously; no further action necessary
+                identifierWithCaret = "Implementation$$";
```

Comment:
is go-to-impl even synchronous when under the time limit?  I think it's async for both.  Just that if the async portion takes under a certain time it navigates, else it pops up the dialog.

Reply:
yup:

```
    private async Task ExecuteCommandWorkerAsync(
        Document document,
        int position,
        CancellationTokenSource cancellationTokenSource)
    {
        // Switch to the BG immediately so we can keep as much work off the UI thread.
        await TaskScheduler.Default;
```

So, in order to do this, i think you'll need to at least wait the `GoToImplementation` test listener.  than after that you can check teh state of the editor.

Reply:
We already are waiting for this:
https://github.com/dotnet/roslyn/blob/37e7ae25a0d16892d1df9a9a7a08f9d3ed40e53d/src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorInProcess.cs#L910-L912

Reply:
ah ok.  thanks!